---
title: "Testing for Genetic Variability in Geographic Space"
author: "Mary O'Neill"
date: "April 27, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#setwd("C:/Users/Mary/PepLab/data/Phylogeography/mantel/")
require(ape)
require(adegenet)
require(ade4)
#require(mmod)
require(vegan)
#require(fields)
#require(ecodist)
#require(hierfstat)

```

## Distance Matrices

In order to perform a Mantel test I need a genetic distance matrix and a geographic distance Matrix. I have downloaded two distance matrices generated by Alex Zarley: one is the the distance between isolates from their sample locations and the other is the distance along the trading posts that they were asigned to. 

```{r geo}

#Distance matrix for trade routes
trade <- read.table("C:/Users/Mary/PepLab/data/Phylogeography/mantel/160511_Route_Dist.txt", header=T, row.names=1, sep='\t')
dm.trade <- as.matrix(trade)

#Distance matrix for TB samples
tbdist <- read.table("C:/Users/Mary/PepLab/data/Phylogeography/mantel/160511_Geodesic_Dist.txt", header=T, row.names=1, sep='\t')
tbdist[is.na(tbdist)] <- 0
dm.tb <- as.matrix(tbdist)



```

## Genetic Distances

I use the APE package to read in the SNP alignment.

```{r dna}

#The meta data for the isolates
meta <- read.table("C:/Users/Mary/PepLab/data/Phylogeography/mantel/160223_passing2.meta", header=F, sep='\t') #read the file
row.names(meta) <- meta$V1
names(meta) <- c("name", "accession", "country", "subregion", "iso2", "whoreg", "lin", "lat", "long", "permissing")

#Read in the SNP alignment
Mtb <- read.FASTA("C:/Users/Mary/PepLab/data/Phylogeography/mantel/160223_snps_50per_recoded.fasta") #read the fasta
#pops <- sapply(strsplit(labels(Mtb), "_"), "[[", 2) #extract pop from isolate name
#Mtb2 <- DNAbin2genind(Mtb, pops) #convert DNAbin to genind

#popcag <- genind2genpop(Mtb2) #convert from genind to genpop object

#nei <- dist.genpop(popcag, method=1)

#Fst calculation takes to long and freezes up computer???
#Fst <- pairwise.fst(Mtb2, pop=NULL, res.type="matrix")


genetic.dists = dist.dna(Mtb) #pairwise distance matrix computation
gm <- as.matrix(genetic.dists)
rownames(gm) <- sapply(strsplit(rownames(gm), "_"), "[[", 1)
colnames(gm) <- sapply(strsplit(colnames(gm), "_"), "[[", 1)
  
gm <- gm[rownames(gm) %in% rownames(dm.trade), colnames(gm) %in% colnames(dm.trade)]


samps <- intersect(rownames(dm.trade), rownames(dm.tb))
setdiff(samps, sapply(strsplit(labels(Mtb), "_"), "[[", 1))
setdiff(samps, meta$name)

#The Kiribati samples
setdiff(sapply(strsplit(labels(Mtb), "_"), "[[", 1), samps)

```

##Mantel Tests
###Whole Dataset

```{r mantel}

distsamps <- rownames(dm.trade)
gensamps <- rownames(gm)
samps <- intersect(distsamps, gensamps)

dm.a <- dm.trade[rownames(dm.trade) %in% samps, colnames(dm.trade) %in% samps]
gm.a <- gm[rownames(gm) %in% samps, colnames(gm) %in% samps]

sdm.a <- dm.a[order(rownames(dm.a)), order(colnames(dm.a))]
sgm.a <- gm.a[order(rownames(gm.a)), order(colnames(gm.a))]


mantel.test(sdm.a, sgm.a)
mantel(sgm.a,sdm.a)
mantel.rtest(as.dist(sdm.a),as.dist(sgm.a))

```

###Lineage 1

```{r lin1}
#subset matrices by lineage 1
lin1 = as.vector(subset(meta, lin == 1, name))

dm.1 <- dm.trade[rownames(dm.trade) %in% lin1$name, colnames(dm.trade) %in% lin1$name]
gm.1 <- gm[rownames(gm) %in% lin1$name, colnames(gm) %in% lin1$name]

sdm.1 <- dm.1[order(rownames(dm.1)), order(colnames(dm.1))]
sgm.1 <- gm.1[order(rownames(gm.1)), order(colnames(gm.1))]


mantel.test(sdm.1, sgm.1)

mantel(sgm.1,sdm.1)

mantel.rtest(as.dist(sdm.1),as.dist(sgm.1))
```

###Lineage 2

```{r lin2}
#subset matrices by lineage 2
lin2 = as.vector(subset(meta, lin == 2, name))

dm.2 <- dm.trade[rownames(dm.trade) %in% lin2$name, colnames(dm.trade) %in% lin2$name]
gm.2 <- gm[rownames(gm) %in% lin2$name, colnames(gm) %in% lin2$name]

sdm.2 <- dm.2[order(rownames(dm.2)), order(colnames(dm.2))]
sgm.2 <- gm.2[order(rownames(gm.2)), order(colnames(gm.2))]


mantel.test(sdm.2, sgm.2)
mantel(sgm.2,sdm.2)
mantel.rtest(as.dist(sdm.2),as.dist(sgm.2))
```

###Lineage 3

```{r lin3}
#subset matrices by lineage 3
lin3 = as.vector(subset(meta, lin == 3, name))

dm.3 <- dm.trade[rownames(dm.trade) %in% lin3$name, colnames(dm.trade) %in% lin3$name]
gm.3 <- gm[rownames(gm) %in% lin3$name, colnames(gm) %in% lin3$name]

sdm.3 <- dm.3[order(rownames(dm.3)), order(colnames(dm.3))]
sgm.3 <- gm.3[order(rownames(gm.3)), order(colnames(gm.3))]



mantel.test(sdm.3, sgm.3)
mantel(sgm.3,sdm.3)
mantel.rtest(as.dist(sdm.3),as.dist(sgm.3))

```


###Lineage 4

```{r lin4}
#subset matrices by lineage 4
lin4 = as.vector(subset(meta, lin == 4, name))

dm.4 <- dm.trade[rownames(dm.trade) %in% lin4$name, colnames(dm.trade) %in% lin4$name]
gm.4 <- gm[rownames(gm) %in% lin4$name, colnames(gm) %in% lin4$name]

sdm.4 <- dm.4[order(rownames(dm.4)), order(colnames(dm.4))]
sgm.4 <- gm.4[order(rownames(gm.4)), order(colnames(gm.4))]



mantel.test(sdm.4, sgm.4)
mantel(sgm.4,sdm.4)
mantel.rtest(as.dist(sdm.4),as.dist(sgm.4))
```

###Lineage 5

```{r lin5}
#subset matrices by lineage 5
lin5 = as.vector(subset(meta, lin == 5, name))

dm.5 <- dm.trade[rownames(dm.trade) %in% lin5$name, colnames(dm.trade) %in% lin5$name]
gm.5 <- gm[rownames(gm) %in% lin5$name, colnames(gm) %in% lin5$name]

sdm.5 <- dm.5[order(rownames(dm.5)), order(colnames(dm.5))]
sgm.5 <- gm.5[order(rownames(gm.5)), order(colnames(gm.5))]



#mantel.test(sdm.5, sgm.5)
#mantel(sgm.5,sdm.5)
#mantel.rtest(as.dist(sdm.5),as.dist(sgm.5))
```

###Lineage 6

```{r lin6}
#subset matrices by lineage 6
lin6 = as.vector(subset(meta, lin == 6, name))

dm.6 <- dm.trade[rownames(dm.trade) %in% lin6$name, colnames(dm.trade) %in% lin6$name]
gm.6 <- gm[rownames(gm) %in% lin6$name, colnames(gm) %in% lin6$name]

sdm.6 <- dm.6[order(rownames(dm.6)), order(colnames(dm.6))]
sgm.6 <- gm.6[order(rownames(gm.6)), order(colnames(gm.6))]



mantel.test(sdm.6, sgm.6)
mantel(sgm.6,sdm.6)
mantel.rtest(as.dist(sdm.6),as.dist(sgm.6))
```


###Lineage 7

```{r lin7}
#subset matrices by lineage 7
lin7 = as.vector(subset(meta, lin == 7, name))

dm.7 <- dm.trade[rownames(dm.trade) %in% lin7$name, colnames(dm.trade) %in% lin7$name]
gm.7 <- gm[rownames(gm) %in% lin7$name, colnames(gm) %in% lin7$name]

sdm.7 <- dm.7[order(rownames(dm.7)), order(colnames(dm.7))]
sgm.7 <- gm.7[order(rownames(gm.7)), order(colnames(gm.7))]



#mantel.test(sdm.7, sgm.7)
#mantel(sgm.7,sdm.7)
#mantel.rtest(as.dist(sdm.7),as.dist(sgm.7))
```
