---
title: "Quality Control - Phylogeography"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

require("ggplot2")


```

This is meant to look for outlier samples in the Phylogeography dataset. I am going to start by reading in the output from a script I wrote to grab summary statistics for any number of samples from the "genome_results.txt" file output by Qualimap bamqc (for reference-guided 'RGA' samples) and Quast (for de novo samples). My scripts can be found on github (https://github.com/ONeillMB1/alignmentQC/). Information about the program Qualimap can be found on its website (http://qualimap.bioinfo.cipf.es). Information about the program Quast can be found on its website (http://quast.bioinf.spbau.ru/manual.html)

Lets start by reading in the datasets, combining them, and having a look at the final dataframe's structure. A note about the 'Type': the dataset is comprised of an assortment of summary statistics about various sequence read archives (SRA) that have been mapped to H37Rv with the MEM algorithm, as well as some de novo assemblies. For the SRAs, they can be broken down further into sample for which there is only one SRA, designated 'solo', or were there exists greater than 1 SRA for a given biosample, designated ('multi'). (The Beijing study had multiple sequencing runs per sample in some cases.)

```{r readFile}
#specify paths to input files
soloFile <- "C:/Users/Mary/PepLab/data/Phylogeography/QualityControl/new_all_RGA_solo_collatedBamQC.txt"
multiFile <- "C:/Users/Mary/PepLab/data/Phylogeography/QualityControl/all_RGA_multi_collatedBamQC.txt"

denovoFile <- "C:/Users/Mary/PepLab/data/Phylogeography/QualityControl/denovo_collatedQuast.txt"

#read in the files
solo <- read.csv(soloFile, header=T, sep='\t', na.strings="NA")
multi <- read.csv(multiFile, header=T, sep='\t', na.strings="NA")
denovo <- read.csv(denovoFile, header=T, sep='\t', na.strings="NA")

#Quast only outputs the number of indels, so make a column in the solo and multi by adding the number of insertions and number of deletions
solo$indels <- solo$numIns + solo$numDel
multi$indels <- multi$numIns + multi$numDel

#Create a column specifying the sample type: solo, multi, or denovo
solo$Type <- "solo"
multi$Type <- "multi"
denovo$Type <- "denovo"

#Merge the files into one dataframe
dat <- rbind(solo, multi, denovo)

#Force type into a factor
dat$Type <- as.factor(dat$Type)

str(dat)
```

There are `r length(solo$Sample)` 'solo' samples in the dataset.There are `r length(multi$Sample)` 'multi' samples in the dataset. (Multi sample meaning that there are more than on SRA runs for the biosample.) There are `r length(denovo$Sample)` 'denovo' samples in the dataset.


##Mean Mapping Quality

```{r MQ}

MQ <- ggplot(dat, aes(Type, meanMQ)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  scale_y_continuous(limits = c(54,61)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Mean Mapping Quality") 
  
plot(MQ)
```



##Mean Coverage

```{r cov}

cov <- ggplot(dat, aes(Type, meanCov)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  scale_y_log10() +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Mean Coverage") +
  geom_hline(aes(yintercept = 30), color = "red", linetype="dashed") +
  geom_hline(aes(yintercept = 20), color = "blue", linetype="dashed")

  
plot(cov)
```

The dashed red line indicates an average coverage of 30X. Lets look at the samples that have low coverage.

```{r lowCov}

LC <- subset(dat, meanCov < 20, select = c(Sample, meanCov, stdCov, perGen10X, perGen15X, perGen20X, perGen30X, perMapped, GCper, Type))

LC[order(LC$meanCov),]

```

Many of these samples are from the global diversity dataset :( Lets look at some other summary statistics.I think the percentage of the genome covered by at least 10X (perGen10X) is a more important factor. 

##Percent Mapped

```{r pm}

pm <- ggplot(dat, aes(x=Type, y=perMapped)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Percent Mapped") 

  
plot(pm)
```

Yikes.... some of these have a very low percent mapped, likely indicating contamination. Lets look at those samples with less than 80% of their reads mapping to H37Rv.

```{r lowPerMap}

LPM <- subset(dat, perMapped < 80, select = c(Sample, meanCov, stdCov, perGen10X, perGen15X, perGen20X, perGen30X, perMapped, GCper, Type))

LPM[order(LPM$perMapped),]

```


Lets look at the relationship between percent mapped and mean coverage.

```{r PMvCov}

PMvCov <- ggplot(dat, aes(x=perMapped, y=meanCov)) + 
  geom_point(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  scale_y_log10() +
  geom_hline(aes(yintercept = 30), color = "red", linetype="dashed") +
  geom_hline(aes(yintercept = 20), color = "blue", linetype="dashed") +
  xlab("Percent Mapped") +
  ylab("Mean Coverage")
  

  
plot(PMvCov)
```

The red dotted line corresponds to a mean coverage of 30X and the blue a mean coverage of 20X. There appears to be a bit of correlation between samples with a low percentage of total reads mapped and coverage. However, at higher percent mapped reads we see a wide distribution of coverage. This is not surprising given that some samples were sequenced several years ago and others more recently.  

##Duplication Rate

```{r dup}

dup <- ggplot(dat, aes(x=Type, y=dupRate)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Duplication Rate") 

  
plot(dup)
```

I honestly don't know how to interpret the duplication rate. We also mark duplicates in our pipeline, so I am less concerned about this statistic. Still, it is shocking how high this is in many of the samples.As for the de novo samples, Quast reports a "duplication ratio" and I multiplied this value by 100 in attempt to normalize this to the "duplication rate" value calculatie by Qualimap bamqc, however I'm not sure if the two are actually comparable.

##Median Insert Size

```{r ins}

ins <- ggplot(dat, aes(x=Type, y=medIns)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Median Insert Size")
  
plot(ins)
```

The samples with zero are single-ended and thus do not have any insert size. It appears the majority of the samples were 250bp paired-end libraries. This statistic does not matter, it is just reflective of library prep.


##GC Percentage

```{r gc}

gc <- ggplot(dat, aes(x=Type, y=GCper)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("GC Percentage") 

  
plot(gc)
```

I don't know how much flexibility we should allow in terms of GC content for M.tb? Perhaps different lineages have differing gc content? Maybe I can incorporate lineage info into this in the future. It is of note that the denovo samples have a very tight GC distriubtion. Perhaps this is reflective of the majority of them being prepared by the Broad, while the SRAs are from numerous different studies?

For now, lets look at the relationship between percent mapped and GC content.

```{r corr}

corr <- ggplot(dat, aes(x=perMapped, y=GCper, group=Type)) + 
  geom_point(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("perMapped") +
  ylab("GC Percentage")
  

  
plot(corr)
```

It looks like there may be a correlation at lower percent mapped, but those above ~85% mapped seem to have a wide spread. Note that I do not have a percent mapped for denovo samples so they are not plotted despite showing up in the legend. 

##Nucleotide Content

```{r nuc}

require("reshape")
nuc <- dat[,c('Sample','Type', 'A','C','G','T')]
nuc.m <- melt(nuc, id = c('Sample', 'Type'))

nucPlot <- ggplot(nuc.m, aes(x=variable, y=value)) + 
  geom_boxplot(aes(fill = Type), color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("Nucleotide") +
  ylab("Percentage") 

  
plot(nucPlot)
```

The G-C content should be the about the same for each sample, so maybe I can identify some outliers by plotting the G vs C content and T vs A.

```{r CvG}

nuc2 <- ggplot(dat, aes(x=C, y=G)) + 
  geom_point(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("Percentage C") +
  ylab("Percentage G") 

  
plot(nuc2)

```

I see one clear outlier. Lets figure out which it is. 

```{r outlier}

dat$GvC <- dat$G/dat$C

tail(na.omit(dat[order(dat$GvC),]), n = 1)

```

This is sample ERR040131. Interestingly it has a pretty normal GC percentage. It also has an extremely high mean coverage, but the percentage of the genome covered at a minimum of 30X is < 80% ? This strikes me as strange.


Lets move on and do the same for A vs T. 
```{r AvT}

nuc3 <- ggplot(dat, aes(x=A, y=T)) + 
  geom_point(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("Percentage A") +
  ylab("Percentage T") 

  
plot(nuc3)


```

There are two samples that look like clear outliers. 

```{r outlier2}

dat$AvT <- dat$A/dat$T

head(na.omit(dat[order(dat$AvT),]), n = 2)

```

Again, sample ERR040131 shows up. Something weird is going on with this sample. Some kind of strand bias perhaps? It also has a crazy high duplication rate, though admittedly, so did several other samples.

The other outlier in this comparison is ERR234185. It is part of the global Mtb diversity set, but actually has a decent mean coverage and coverage across the genome. 


##Percent Genome Covered at >= X

```{r covPlot}


cov <- dat[,c('Sample', "Type", 'meanCov', 'stdCov', 'perGen10X', 'perGen15X', 'perGen20X', 'perGen30X')]
cov.m <- melt(cov, id = c('Sample', 'meanCov', 'stdCov', "Type"))

covPlot <- ggplot(cov.m, aes(x=variable, y=value)) + 
  geom_boxplot(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("X Coverage") +
  ylab("Percent Genome Covered") 
  

  
plot(covPlot)
```

I don't know what a good threshold should be. Lets zoom in a bit.

```{r covPlot2}


covPlot2 <- ggplot(cov.m, aes(x=variable, y=value)) + 
  geom_boxplot(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("X Coverage") +
  ylab("Percent Genome Covered") +
  scale_y_continuous(limits=c(75,100))
  
  

  
plot(covPlot2)
```

It seems like there is a natural break at about 90%, 85%, and 75%. Could these be based on lineages? Again, I would like to incorporate that data. 

Also, the fraction of the genome covered for denovo samples are not directly comparable to those reported for the SRA samples (it is reported the same across all coverages). This value only reclects the portion of the H37Rv genome fraction found in the assembly. 

How about we look at the relationship between the percent genome covered at a given fraction vs the mean coverage. 

```{r meanCov_perCov}

MCvPC <- ggplot(cov.m, aes(x=meanCov, y=value)) + 
  facet_wrap(~variable) +
  geom_point(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("Mean Coverage") +
  scale_x_log10() +
  ylab("Fraction Genome Covered") 

  
plot(MCvPC)
```

Again we do not have 'mean coverage' info for the denovo samples so they are not plotted. Some of the 'multi' samples have a low fraction of the genome covered, however, for each of the samples in this group there are multiple runs. 

##Error Rate

```{r err}

err <- ggplot(dat, aes(x=Type, y=errRate)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Error Rate") 

  
plot(err)
```

Honestly not really sure how to interpret this one either. There does seem to be a stark division among the 'solo' runs. Perhaps this is reflective of different studies?

##Mismatches

```{r misMat}

misMat <- ggplot(dat, aes(x=Type, y=misMat)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Mismatches") 

  
plot(misMat)
```

I don't think the 'mismatches' reported by Quast (denovo) are comparable to the 'mismatches' output by Qualimap. I would like to see if the number of mismatches correlates with lineage. 

For now, lets see if it correlates with mean coverage...


```{r MisVsCov}

MisVsCov <- ggplot(dat, aes(x=meanCov, y=misMat, group=Type)) + 
  geom_point(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("meanCov") +
  ylab("Mismatches")
  

  
plot(MisVsCov)

```

At first pass, this is a bit alarming. However, I believe what this is reporting is simply the number of bases differing from the reference AND not necessarily that it will be called a SNP. You can imagine that if the error rate is constant, as you increase coverage you are bound to have more incorrecltly called bases, but our power to detect such even will increase. So this probably isn't the best stat to look at. Or I should normalize by the total number of bases or something. 

##Indels

```{r indels}

indels <- ggplot(dat, aes(x=Type, y=indels)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Indels") 

  
plot(indels)
```

Here is another statistic that I do not think are comparable in terms of Quast vs Qualimap. I do not know with what respect Quast is calling indels. This again could be affected by lineage.

##Homopolymer Indels

```{r hom}

hom <- ggplot(dat, aes(x=Type, y=homInd)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Homopolymer Indels (Percentage)") 

  
plot(hom)
```

Lets look at the relationship between percent homopolymer indels and absolute number of indels.
```{r PerVsAbs}

PerVsAbs <- ggplot(dat, aes(x=indels, y=homInd, group=Type)) + 
  geom_point(aes(colour=Type)) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("indels") +
  ylab("Percent Homopolymer")
  

  
plot(PerVsAbs)

```

That does not seem to explain anything. 


##N's

Finally, lets look at a denovo specific statistic - the number of "N".

```{r N}

N <- ggplot(dat, aes(x=Type, y=N)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  xlab("") +
  ylab("Number of N's")
  

  
plot(N)
```

There certainly seem to be a few outliers in this department. Lets see which samples they are. 

```{r Noutier}

tail(dat[order(dat$N),], n = 6)

```

I would like to reference the tree and see if the samples identified as outliers here are odd balls on the tree.