Mantel Tests for Phylogeography
========================================================

#This is an R Markdown document. Markdown is a simple formatting syntax for authoring web pages (click the **Help** toolbar button for more details on using R Markdown).

#When you click the **Knit HTML** button a web page will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

In order to perform a Mantel test I need a genetic distance matrix and a geographic distance matrix. As the latitude and longitude coordinates for many samples may change once we decide on the best method of assigning a location, I thought it best to keep coordinates in a text file. This way we can also have the coordinates of the place of birth, as well as add in coordinates for the most likely silk road post. Using a bit of R code, I can transform latitude and longitude coordinates into a distance matrix.

```{r}

setwd("C:/Users/Mary/Desktop/phylo")
meta <- read.table("150924_lineages_meta_recode.txt", header=F, sep='\t') #read the file
row.names(meta) <- meta$V1
names(meta) <- c("name", "samp", "lin", "study", "study.samp", "country", "pre", "lat", "long") #give headers
cd <- meta[,c('lat','long')]
country.dists <- dist(cd) #distance matrix computation
dm <- as.matrix(country.dists)

```

Note: I still need to subset and order the geographic distance matrix to match that of the genetic distance matrix that will be computed. I will need to learn how to keep the row.names for help ordering/matching these two matrices. 


To generate the genetic distance matrix I will use the R package 'Ape'.

```{r}
require(ape)
Mtb <- read.FASTA("masked_merged_snps.fasta") #read the fasta
genetic.dists = dist.dna(Mtb) #pairwise distance matrix computation
gm <- as.matrix(genetic.dists)

```


Playing around with the pcoa function in 'Ape'.

```{r}
res <- pcoa(genetic.dists) #principal coordinate analysis
biplot(res) #simple biplot

# 3D Scatterplot
require(scatterplot3d)
attach(res)
scatterplot3d(res$vectors[,1],res$vectors[,2],res$vectors[,3], main="") #3D plot!

```




```{r}
require(vegan)
require(ade4)
require(ape)
require(cba)

#dm.X <- dm[rownames(dm) %in% linX, colnames(dm) %in% linX]

#subset matrices by lineage 1
lin1 = as.vector(subset(meta, lin == 1, name))

dm.1 <- dm[rownames(dm) %in% lin1$name, colnames(dm) %in% lin1$name]
gm.1 <- gm[rownames(gm) %in% lin1$name, colnames(gm) %in% lin1$name]

sdm.1 <- dm.1[order(rownames(dm.1)), order(colnames(dm.1))]
sgm.1 <- gm.1[order(rownames(gm.1)), order(colnames(gm.1))]


mantel.test(sdm.1, sgm.1)

mantel(sgm.1,sdm.1)

mantel.rtest(as.dist(sdm.1),as.dist(sgm.1))

#subset matrices by lineage 2
lin2 = as.vector(subset(meta, lin == 2, name))

dm.2 <- dm[rownames(dm) %in% lin2$name, colnames(dm) %in% lin2$name]
gm.2 <- gm[rownames(gm) %in% lin2$name, colnames(gm) %in% lin2$name]

sdm.2 <- dm.2[order(rownames(dm.2)), order(colnames(dm.2))]
sgm.2 <- gm.2[order(rownames(gm.2)), order(colnames(gm.2))]


mantel.test(sdm.2, sgm.2)
mantel(sgm.2,sdm.2)
mantel.rtest(as.dist(sdm.3),as.dist(sgm.3))

#subset matrices by lineage 3
lin3 = as.vector(subset(meta, lin == 3, name))

dm.3 <- dm[rownames(dm) %in% lin3$name, colnames(dm) %in% lin3$name]
gm.3 <- gm[rownames(gm) %in% lin3$name, colnames(gm) %in% lin3$name]

sdm.3 <- dm.3[order(rownames(dm.3)), order(colnames(dm.3))]
sgm.3 <- gm.3[order(rownames(gm.3)), order(colnames(gm.3))]



mantel.test(sdm.3, sgm.3)
mantel(sgm.3,sdm.3)
mantel.rtest(as.dist(sdm.3),as.dist(sgm.3))


#subset matrices by lineage 4
lin4 = as.vector(subset(meta, lin == 4, name))

dm.4 <- dm[rownames(dm) %in% lin4$name, colnames(dm) %in% lin4$name]
gm.4 <- gm[rownames(gm) %in% lin4$name, colnames(gm) %in% lin4$name]

sdm.4 <- dm.4[order(rownames(dm.4)), order(colnames(dm.4))]
sgm.4 <- gm.4[order(rownames(gm.4)), order(colnames(gm.4))]



mantel.test(sdm.4, sgm.4)
mantel(sgm.4,sdm.4)
mantel.rtest(as.dist(sdm.4),as.dist(sgm.4))



#subset matrices by lineage 5
lin5 = as.vector(subset(meta, lin == 5, name))

dm.5 <- dm[rownames(dm) %in% lin5$name, colnames(dm) %in% lin5$name]
gm.5 <- gm[rownames(gm) %in% lin5$name, colnames(gm) %in% lin5$name]

sdm.5 <- dm.5[order(rownames(dm.5)), order(colnames(dm.5))]
sgm.5 <- gm.5[order(rownames(gm.5)), order(colnames(gm.5))]



mantel.test(sdm.5, sgm.5)
mantel(sgm.5,sdm.5)
mantel.rtest(as.dist(sdm.5),as.dist(sgm.5))

#subset matrices by lineage 6
lin6 = as.vector(subset(meta, lin == 6, name))

dm.6 <- dm[rownames(dm) %in% lin6$name, colnames(dm) %in% lin6$name]
gm.6 <- gm[rownames(gm) %in% lin6$name, colnames(gm) %in% lin6$name]

sdm.6 <- dm.6[order(rownames(dm.6)), order(colnames(dm.6))]
sgm.6 <- gm.6[order(rownames(gm.6)), order(colnames(gm.6))]



mantel.test(sdm.6, sgm.6)
mantel(sgm.6,sdm.6)
mantel.rtest(as.dist(sdm.6),as.dist(sgm.6))

#subset matrices by lineage 7
lin7 = as.vector(subset(meta, lin == 7, name))

dm.7 <- dm[rownames(dm) %in% lin7$name, colnames(dm) %in% lin7$name]
gm.7 <- gm[rownames(gm) %in% lin7$name, colnames(gm) %in% lin7$name]

sdm.7 <- dm.7[order(rownames(dm.7)), order(colnames(dm.7))]
sgm.7 <- gm.7[order(rownames(gm.7)), order(colnames(gm.7))]



mantel.test(sdm.7, sgm.7)
mantel(sgm.7,sdm.7)
mantel.rtest(as.dist(sdm.7),as.dist(sgm.7))

```



For the source-sink model, I need a way to calculate nucleotide diversity. Previously I used the PopGenome package. I also want to explore the Pegas package.

```{r}
require(pegas)
require(adegenet)


```

